### ISSUE TITLE
Infinite loop in ECDSA sign

### ISSUE THREAD

**Issue by zanebeckwith** [milagro-crypto/milagro-crypto-c#280]
In the ECDSA sign function `ECP_ZZZ_SP_DSA`, it looks like it could enter an infinite loop if the caller provides the ephemeral randomness `k` and either component of the signature is `0`.

Could the function return some error code to indicate that it needs to be called again with a new `k` in this case instead?

**Comment by mcarrickscott** [comment]
Yes, this should be fixed (although the chances of it happening are
vanishingly  small).

The reason that k might be provided externally is because versions of ecdsa
have been advocated where no external source of "randonness" is used, but
rather the needed "randomness" is rather derived by hashing the message to
be signed.


Mike

**Comment by zanebeckwith** [comment]
Right, the probability of either signature component being 0 is about `1/q` (`q` the order of the group), right? Not something that's keeping me up at night, but I thought I'd raise the issue with you all.

My application actually isn't using deterministic ECDSA signatures; it specifies the `k` manually rather than using the AMCL csprng because it already has its own source of randomness. Is there a reason to be wary of this (other than the usual caveat of making sure you trust your random numbers)?

**Comment by zanebeckwith** [comment]
Also, thanks for the quick response!

**Comment by mcarrickscott** [comment]
No, no reason... :)

Mike

**Comment by zanebeckwith** [comment]
Thanks Mike!
